stdlib/email/message.pyi

```diff
@@ -23,16 +33,40 @@ class Message:
     def is_multipart(self) -> bool: ...
     def set_unixfrom(self, unixfrom: str) -> None: ...
     def get_unixfrom(self) -> str | None: ...
-    def attach(self, payload: Message) -> None: ...
-    def get_payload(self, i: int | None = None, decode: bool = False) -> Any: ...  # returns _PayloadType | None
-    def set_payload(self, payload: _PayloadType, charset: _CharsetType = None) -> None: ...
+    def attach(self, payload: _PayloadType) -> None: ...
+    # `i: int` without a multipart payload results in an error
+    # `| Any`: can be None for cleared or unset payload, but annoying to check
+    @overload  # multipart
+    def get_payload(self, i: int, decode: Literal[True]) -> None: ...
+    @overload  # multipart
+    def get_payload(self, i: int, decode: Literal[False] = False) -> _PayloadType | Any: ...
+    @overload  # either
+    def get_payload(self, i: None = None, decode: Literal[False] = False) -> _PayloadType | _MultipartPayloadType | Any: ...
+    @overload  # not multipart
+    def get_payload(self, i: None = None, *, decode: Literal[True]) -> _EncodedPayloadType | Any: ...
+    @overload  # not multipart, IDEM but w/o kwarg
+    def get_payload(self, i: None, decode: Literal[True]) -> _EncodedPayloadType | Any: ...
+    # If `charset=None` and payload supports both `encode` AND `decode`, then an invalid payload could be passed, but this is unlikely
+    # Not[_SupportsEncodeToPayload]
+    @overload
+    def set_payload(
+        self, payload: _SupportsDecodeToPayload | _PayloadType | _MultipartPayloadType, charset: None = None
+    ) -> None: ...
+    @overload
+    def set_payload(
+        self,
+        payload: _SupportsEncodeToPayload | _SupportsDecodeToPayload | _PayloadType | _MultipartPayloadType,
+        charset: Charset | str,
+    ) -> None: ...
     def set_charset(self, charset: _CharsetType) -> None: ...
     def get_charset(self) -> _CharsetType: ...
     def __len__(self) -> int: ...
     def __contains__(self, name: str) -> bool: ...
     def __iter__(self) -> Iterator[str]: ...
-    def __getitem__(self, name: str) -> _HeaderType: ...
-    def __setitem__(self, name: str, val: _HeaderType) -> None: ...
+    # Same as `get` with `failobj=None`, but with the expectation that it won't return None in most scenarios
+    # This is important for protocols using __getitem__, like SupportsKeysAndGetItem
+    def __getitem__(self, name: str) -> _HeaderType | Any: ...  # AnyOf[_HeaderType, None]
```

Comment:
`_HeaderType` is already just a type alias to `Any`, no?

Reply:
Yes, but it's more about the semantics of what to expect for subclasses. And the assumption that *eventually* something will come in the type system to improve the union.

With #11094 this would be something like `_HeaderType | MaybeNone`
Which would preferably evaluate to something akin to `AnyOf[str, Header, None]` if the type system allowed us.

`HttpMessage`'s overrides should just copy `Message` and replace `_HeaderType` with str. Which would be: `def __getitem__(self, name: str) -> str | MaybeNone: ...` ("Any trick" alias name pending).

@srittau mentioned wanting an actual `str | None` union there. But I am hesitant as it leads to false positives with protocols using `__getitem__` that should otherwise be safe (see the `urllib3` use above of `dict(...)` where the parameter is a `SupportsKeysAndGetItem`.

Usage of the Any trick in `HttpMessage.__getitem__` does seem to bypass that kind of false-positives (quick test over at https://github.com/Avasam/typeshed/pull/37/files#diff-ced02b01b42b244e9b7bc2e12ef7a215bde89a29072c99ade903f58fb83a3bc7R68 )

---

I might combine this PR and #11114 to centralize discussion, I thought the `str` overrides in `HttpMessage` would be a more contentious point and I took a habit of splitting PRs in smaller chunks ^^

Reply:
Hmm. But this just seems confusing to me -- I worry that others will come along and ask the question I just asked. Maybe just making the comment a bit longer would be better here?

```suggestion
    # Morally, the return type should be `AnyOf[_HeaderType, None]`,
    # which we could spell as `_HeaderType | Any`,
    # *but* `_HeaderType` itself is currently an alias to `Any`...
    def __getitem__(self, name: str) -> _HeaderType: ...
```

---

> I took a habit of splitting PRs in smaller chunks ^^

Splitting PRs into smaller chunks is definitely Good!!

Reply:
Clear and explicit. I like your suggestion